### PR TITLE
fix: create unique endpoint configuration name

### PR DESCRIPTION
**Pull request by ahsan-z-khan**
*Issue #, if available:*
[TT](https://t.corp.amazon.com/P46903451/overview)
*Description of changes:*
Added variable to create unique endpoint configuration name.
*Testing done:*
Unit tests
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/session.py [hunk 1]
```diff
@@ -3463,8 +3464,11 @@ def endpoint_from_production_variants(
         Returns:
             str: The name of the created ``Endpoint``.
         """
+        endpoint_config_name = unique_name_from_base(name)
         if not _deployment_entity_exists(
```
Comment (icywang86rui) [inline src/sagemaker/session.py:3468]: Do we still need to check here? We are using a unique name.

**Review by icywang86rui — COMMENTED** [review]
I am not sure we chose to reuse the endpoint_config. There might be a limit of number of endpoint configs you can have. I am also not sure if we clean up endpoint configs in our tests. Could you please confirm before shipping this?

File: src/sagemaker/session.py (hunk 1 above)
Comment (ahsan-z-khan) [reply src/sagemaker/session.py:3468]: > I am not sure we chose to reuse the endpoint_config. There might be a limit of number of endpoint configs you can have. I am also not sure if we clean up endpoint configs in our tests. Could you please confirm before shipping this?

seems like we don't delete the endpoints in test.

Comment (ahsan-z-khan) [reply src/sagemaker/session.py:3468]: deleted it

**Comment by ahsan-z-khan** [comment]
> I am not sure we chose to reuse the endpoint_config. There might be a limit of number of endpoint configs you can have. I am also not sure if we clean up endpoint configs in our tests. Could you please confirm before shipping this?

seems like we don't delete the endpoints in test. I've added a function to check endpoint config.

File: src/sagemaker/session.py [hunk 2]
```diff
@@ -4196,6 +4202,21 @@ def _deployment_entity_exists(describe_fn):
...
+def _endpoint_config_name_exists(describe_fn):
+    """Checks if similar endpoint configuration name exists."""
+    try:
+        describe_fn()
+        return True
+    except ClientError as ce:
+        error_code = ce.response["Error"]["Code"]
+        if not (
+            error_code == "ValidationException"
+            and "Similar endpoint configuration name exists" in ce.response["Error"]["Message"]
```
Comment (icywang86rui) [inline src/sagemaker/session.py:4214]: Is this the exact error message? Most List* apis just returns empty list if there is nothing matching.

File: src/sagemaker/session.py [hunk 3]
```diff
@@ -4196,6 +4202,21 @@ def _deployment_entity_exists(describe_fn):
         return False
 
 
+def _endpoint_config_name_exists(describe_fn):
```
Comment (icywang86rui) [inline src/sagemaker/session.py:4205]: This function is 90% duplicated from _deployment_entity_exists. We should eliminate the duplication.

**Pull request closed without merging**